### PR TITLE
Refactor generic http client

### PR DESCRIPTION
**Pull request by nev-r**
turns all HttpClient generation into generic wrappers that i would like to split off into a separate module after a bit

throws all bungie non-success codes as a new BungieError type, then handles them with dim error messages upstream as errors catching, instead of response handling

### CODE REVIEW

File: src/app/bungie-api/bungie-service-helper.ts [hunk 1]
```diff
@@ -36,226 +41,181 @@ const notifyTimeout = _.throttle(
...
+  )
+);
 
 /** Generate an error with a bit more info */
-export function error(message: string, errorCode: PlatformErrorCodes): DimError {
+export function dimError(message: string, errorCode: PlatformErrorCodes): DimError {
   const error: DimError = new Error(message);
   error.code = errorCode;
   return error;
 }
```
Comment (bhollis) [inline src/app/bungie-api/bungie-service-helper.ts:75]: I sorta feel like we should break DimError out into its own thing (a real Error subclass) and do things like pack untranslated error codes and such in there. It'd make sentry a lot easier.

File: src/app/bungie-api/bungie-service-helper.ts [hunk 2]
```diff
@@ -36,226 +41,181 @@ const notifyTimeout = _.throttle(
...
   }
 
-  let data: ServerResponse<any> | undefined;
-  try {
-    data = await response.json();
-  } catch (e) {
-    if (e instanceof SyntaxError) {
-      console.error('Error parsing Bungie.net response', e);
-      throw new Error(t('BungieService.Difficulties'));
+  if (error instanceof HttpStatusError) {
```
Comment (bhollis) [inline src/app/bungie-api/bungie-service-helper.ts:111]: Huh, I didn't know about this

File: src/app/bungie-api/bungie-service-helper.ts [hunk 3]
```diff
@@ -36,226 +41,181 @@ const notifyTimeout = _.throttle(
...
+      throw new Error(
+        navigator.onLine
+          ? t('BungieService.NotConnectedOrBlocked')
+          : t('BungieService.NotConnected')
+      );
+    }
+
+    // Token expired and other auth maladies
+    if (error.status === 401 || error.status === 403) {
+      goToLoginPage();
```
Comment (bhollis) [inline src/app/bungie-api/bungie-service-helper.ts:123]: Note for the future, I'd love to extract this somehow, but I haven't quite figured out what the right pattern would be.

File: src/app/bungie-api/http-client.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+import { HttpClientConfig, HttpClient } from 'bungie-api-ts/http';
+
+import { stringify } from 'querystring';
+import { delay } from 'app/utils/util';
+import _ from 'lodash';
+
+/**
+ * an error indicating a non-200 response code
+ */
+export class HttpStatusError extends Error {
```
Comment (bhollis) [inline src/app/bungie-api/http-client.ts:11]: I remember that at some point this didn't work, but maybe modern TypeScript can handle it, or the fact that we're not transpiling classes makes it work (I assume we're not transpiling classes anymore?)

File: src/app/bungie-api/http-client.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+// FETCH UTILS
+//
+
+/**
+ * returns a fetch-like that will abort the request after some time
+ *
+ * @param fetchFunction use this function to make the request
+ * @param timeout abort request after this many milliseconds
+ */
+export function createFetchWithTimeout(fetchFunction: typeof fetch, timeout: number): typeof fetch {
```
Comment (bhollis) [inline src/app/bungie-api/http-client.ts:85]: Love the composition of these capabilities

File: src/app/bungie-api/http-client.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    const fetchOptions = new Request(url, {
+      method: config.method,
+      body: config.body ? JSON.stringify(config.body) : undefined,
+      headers: config.body
+        ? {
+            'X-API-Key': apiKey,
+            'Content-Type': 'application/json',
+          }
+        : {
+            'X-API-Key': apiKey,
```
Comment (bhollis) [inline src/app/bungie-api/http-client.ts:159]: Did we ever figure out whether there were some calls that didn't need an API key?

File: src/app/bungie-api/http-client.ts (hunk 4 above)
Comment (nev-r) [reply src/app/bungie-api/http-client.ts:11]: not 100% sure what part wouldn't work but i would love help making sure something's not overlooked. it worked ok in testing when i spoofed http and api error codes

File: src/app/bungie-api/http-client.ts (hunk 6 above)
Comment (nev-r) [reply src/app/bungie-api/http-client.ts:159]: a handful seem to work ok like global settings endpoint, or usually /Manifest/ but i tihnk that may just be caching & nothing doesn't... not.. not work..? with an api key, except the manifest json & its components

File: src/app/bungie-api/bungie-service-helper.ts (hunk 1 above)
Comment (nev-r) [reply src/app/bungie-api/bungie-service-helper.ts:75]: this is partly but i need to merge one thing at a time so diffs are comprehensible. dimerror stretches all over the codebase

File: src/app/bungie-api/http-client.ts (hunk 4 above)
Comment (bhollis) [reply src/app/bungie-api/http-client.ts:8]: If we're using classes directly without transforming them we should be OK. The Error constructor used to return a different instance that broke calls to "super"

**Review by bhollis — APPROVED** [review]

File: src/app/bungie-api/destiny2-api.ts [hunk 7]
```diff
@@ -236,7 +236,7 @@ export async function transfer(
     stackSize: amount || item.amount,
     transferToVault: store.isVault,
   };
-
+  console.log(request);
```
Comment (delphiactual) [inline src/app/bungie-api/destiny2-api.ts:239]: stray?

Comment (nev-r) [reply src/app/bungie-api/destiny2-api.ts:239]: thanks

**Pull request merged**